File: builtins/echo/echo.go
// Unless explicitly stated otherwise all files in this repository are licensed
// under the Apache License Version 2.0.
// This product includes software developed at Datadog (https://www.datadoghq.com/).
// Copyright 2026-present Datadog, Inc.

// Package echo implements the echo builtin command.
//
// echo — write arguments to standard output
//
// Usage: echo [-neE] [ARG]...
//
// Write each ARG to standard output, separated by a single space,
// followed by a newline.
//
// Flags (bash-compatible, not POSIX getopt):
//
//	-n  Do not output a trailing newline.
//	-e  Enable interpretation of backslash escape sequences.
//	-E  Disable interpretation of backslash escapes (default).
//
// Flags can be combined (e.g. -ne). Only leading arguments matching
// -[neE]+ are treated as flags; once a non-flag argument is seen,
// all remaining arguments (including it) are printed as text.
// "--" is NOT treated as an end-of-options separator.
//
// Supported escape sequences (with -e):
//
//	\\    backslash
//	\a    alert (BEL)
//	\b    backspace
//	\c    suppress further output (including trailing newline)
//	\e    escape character (0x1B)
//	\E    escape character (0x1B)
//	\f    form feed
//	\n    newline
//	\r    carriage return
//	\t    horizontal tab
//	\v    vertical tab
//	\0nnn octal value (1 to 3 digits)
//	\xHH  hexadecimal value (1 to 2 digits)
//	\uHHHH    Unicode character (1 to 4 hex digits)
//	\UHHHHHHHH Unicode character (1 to 8 hex digits)
//
// Exit codes:
//
//	0  Always succeeds.
package echo

import (
	"context"
	"strings"

	"github.com/DataDog/rshell/builtins"
)

// Cmd is the echo builtin command descriptor.
var Cmd = builtins.Command{Name: "echo", Description: "write arguments to stdout", MakeFlags: builtins.NoFlags(run)}

func run(_ context.Context, callCtx *builtins.CallContext, args []string) builtins.Result {
	if len(args) > 0 && args[0] == "--help" {
		callCtx.Out("Usage: echo [-neE] [ARG]...\nWrite each ARG to standard output, separated by a single space,\nfollowed by a newline.\n\n  -n  do not output the trailing newline\n  -e  enable interpretation of backslash escapes\n  -E  disable interpretation of backslash escapes (default)\n      --help  display this help and exit\n")
		return builtins.Result{}
	}
	// Parse flags: bash treats leading args matching -[neE]+ as flags.
	// Once a non-matching arg is seen, everything from that point is text.
	var noNewline, escapes bool
	textStart := 0
	for i, arg := range args {
		if !isEchoFlag(arg) {
			break
		}
		textStart = i + 1
		for _, ch := range arg[1:] { // skip leading '-'
			switch ch {
			case 'n':
				noNewline = true
			case 'e':
				escapes = true
			case 'E':
				escapes = false
			}
		}
	}

	textArgs := args[textStart:]

	for i, arg := range textArgs {
		if i > 0 {
			callCtx.Out(" ")
		}
		if escapes {
			s, stop := processEscapes(arg)
			callCtx.Out(s)
			if stop {
				return builtins.Result{}
			}
		} else {
			callCtx.Out(arg)
		}
	}

	if !noNewline {
		callCtx.Out("\n")
	}
	return builtins.Result{}
}

// isEchoFlag returns true if arg looks like a valid echo flag: starts with '-'
// and every subsequent character is one of 'n', 'e', 'E'. Must have at least
// one flag character (i.e. bare "-" is not a flag).
func isEchoFlag(arg string) bool {
	if len(arg) < 2 || arg[0] != '-' {
		return false
	}
	for i := 1; i < len(arg); i++ {
		switch arg[i] {
		case 'n', 'e', 'E':
		default:
			return false
		}
	}
	return true
}

// processEscapes interprets backslash escape sequences in s.
// Returns the processed string and whether \c was encountered (meaning
// all further output including trailing newline should be suppressed).
func processEscapes(s string) (string, bool) {
	var b strings.Builder
	b.Grow(len(s))
	i := 0
	for i < len(s) {
		if s[i] != '\\' || i+1 >= len(s) {
			b.WriteByte(s[i])
			i++
			continue
		}
		// We have a backslash followed by at least one character.
		i++ // skip '\'
		switch s[i] {
		case '\\':
			b.WriteByte('\\')
		case 'a':
			b.WriteByte('\a')
		case 'b':
			b.WriteByte('\b')
		case 'c':
			return b.String(), true
		case 'e', 'E':
			b.WriteByte(0x1B)
		case 'f':
			b.WriteByte('\f')
		case 'n':
			b.WriteByte('\n')
		case 'r':
			b.WriteByte('\r')
		case 't':
			b.WriteByte('\t')
		case 'v':
			b.WriteByte('\v')
		case '0':
			// Octal: \0nnn (up to 3 octal digits after the '0')
			i++
			val, consumed := parseOctal(s[i:], 3)
			i += consumed
			b.WriteByte(byte(val)) // Intentional: wraps values > 255 (matches bash)
			continue
		case 'x':
			// Hex: \xHH (up to 2 hex digits)
			i++
			val, consumed := parseHex(s[i:], 2)
			if consumed == 0 {
				// No valid hex digits: output \x literally
				b.WriteByte('\\')
				b.WriteByte('x')
				continue
			}
			i += consumed
			b.WriteByte(byte(val))
			continue
		case 'u':
			// Unicode: \uHHHH (up to 4 hex digits)
			i++
			val, consumed := parseHex(s[i:], 4)
			if consumed == 0 {
				b.WriteByte('\\')
				b.WriteByte('u')
				continue
			}
			i += consumed
			writeUnicodeCodepoint(&b, int64(val))
			continue
		case 'U':
			// Unicode: \UHHHHHHHH (up to 8 hex digits)
			i++
			val, consumed := parseHex(s[i:], 8)
			if consumed == 0 {
				b.WriteByte('\\')
				b.WriteByte('U')
				continue
			}
			i += consumed
			writeUnicodeCodepoint(&b, val)
			continue
		default:
			// Unrecognized escape: output backslash and the character literally.
			b.WriteByte('\\')
			b.WriteByte(s[i])
		}
		i++
	}
	return b.String(), false
}

// writeUnicodeCodepoint writes a Unicode codepoint to the builder.
// Bash emits raw bytes for surrogates (U+D800-U+DFFF), producing invalid UTF-8,
// and drops values above a certain threshold. We replace surrogates with U+FFFD
// and drop anything beyond U+10FFFF to avoid emitting invalid UTF-8.
func writeUnicodeCodepoint(b *strings.Builder, val int64) {
	if val > 0x10FFFF {
		// Intentional divergence: bash emits raw bytes for values up to 0x7FFFFFFF,
		// but we drop them to avoid producing invalid UTF-8.
		return
	}
	// Go's WriteRune replaces surrogates (U+D800-U+DFFF) with U+FFFD automatically.
	b.WriteRune(rune(val))
}

// parseOctal reads up to maxDigits octal digits from s and returns the
// accumulated value and the number of bytes consumed.
func parseOctal(s string, maxDigits int) (int, int) {
	val := 0
	n := 0
	for n < maxDigits && n < len(s) && s[n] >= '0' && s[n] <= '7' {
		val = val*8 + int(s[n]-'0')
		n++
	}
	return val, n
}

// parseHex reads up to maxDigits hexadecimal digits from s and returns
// the accumulated value and the number of bytes consumed.
// Returns int64 to avoid overflow on 32-bit architectures when parsing
// up to 8 hex digits for \U escape sequences.
func parseHex(s string, maxDigits int) (int64, int) {
	var val int64
	n := 0
	for n < maxDigits && n < len(s) {
		ch := s[n]
		switch {
		case ch >= '0' && ch <= '9':
			val = val*16 + int64(ch-'0')
		case ch >= 'a' && ch <= 'f':
			val = val*16 + int64(ch-'a') + 10
		case ch >= 'A' && ch <= 'F':
			val = val*16 + int64(ch-'A') + 10
		default:
			return val, n
		}
		n++
	}
	return val, n
}

File: rust/crates/rshell-builtins/src/echo.rs
//! `echo` — print arguments separated by spaces, terminated by a newline.
//!
//! Flags: `-n` (no trailing newline), `-e` (interpret backslash escapes),
//! `-E` (disable escapes). Bash convention: leading argv elements that
//! match `-[neE]+` are treated as flags; the first non-flag stops flag
//! parsing.

use rshell_interp::{Builtin, CallCtx};

pub struct Echo;

impl Builtin for Echo {
    fn run(&self, ctx: &mut CallCtx<'_>) -> i32 {
        let mut newline = true;
        let mut interpret_escapes = false;
        let mut idx = 1usize; // skip argv[0]
        while idx < ctx.args.len() {
            let a = ctx.args[idx].as_slice();
            if a.len() < 2 || a[0] != b'-' {
                break;
            }
            // Must consist solely of -, n, e, E to count as a flag bundle.
            if a[1..].iter().all(|c| matches!(c, b'n' | b'e' | b'E')) {
                for c in &a[1..] {
                    match c {
                        b'n' => newline = false,
                        b'e' => interpret_escapes = true,
                        b'E' => interpret_escapes = false,
                        _ => unreachable!(),
                    }
                }
                idx += 1;
            } else {
                break;
            }
        }
        let mut first = true;
        for a in &ctx.args[idx..] {
            if !first {
                if let Err(e) = ctx.stdout.write_all(b" ") {
                    let _ = writeln!(ctx.stderr, "echo: {e}");
                    return 1;
                }
            }
            first = false;
            let payload = if interpret_escapes {
                interpret_backslash_escapes(a.as_slice())
            } else {
                a.to_vec()
            };
            if let Err(e) = ctx.stdout.write_all(&payload) {
                let _ = writeln!(ctx.stderr, "echo: {e}");
                return 1;
            }
        }
        if newline && let Err(e) = ctx.stdout.write_all(b"\n") {
            let _ = writeln!(ctx.stderr, "echo: {e}");
            return 1;
        }
        0
    }
}

fn interpret_backslash_escapes(input: &[u8]) -> Vec<u8> {
    let mut out = Vec::with_capacity(input.len());
    let mut i = 0;
    while i < input.len() {
        if input[i] != b'\\' {
            out.push(input[i]);
            i += 1;
            continue;
        }
        if i + 1 >= input.len() {
            out.push(b'\\');
            i += 1;
            continue;
        }
        match input[i + 1] {
            b'a' => {
                out.push(7);
                i += 2;
            }
            b'b' => {
                out.push(8);
                i += 2;
            }
            b'c' => return out, // suppress further output
            b'e' | b'E' => {
                out.push(0x1b);
                i += 2;
            }
            b'f' => {
                out.push(12);
                i += 2;
            }
            b'n' => {
                out.push(b'\n');
                i += 2;
            }
            b'r' => {
                out.push(b'\r');
                i += 2;
            }
            b't' => {
                out.push(b'\t');
                i += 2;
            }
            b'v' => {
                out.push(11);
                i += 2;
            }
            b'\\' => {
                out.push(b'\\');
                i += 2;
            }
            b'0' => {
                // Octal: up to 3 octal digits after \0.
                i += 2;
                let mut val = 0u32;
                let mut n = 0;
                while n < 3 && i < input.len() && (b'0'..=b'7').contains(&input[i]) {
                    val = val * 8 + (input[i] - b'0') as u32;
                    i += 1;
                    n += 1;
                }
                out.push((val & 0xff) as u8);
            }
            b'x' => {
                // Hex: up to 2 hex digits.
                i += 2;
                let mut val = 0u32;
                let mut n = 0;
                while n < 2 && i < input.len() && input[i].is_ascii_hexdigit() {
                    val = val * 16 + (input[i] as char).to_digit(16).unwrap();
                    i += 1;
                    n += 1;
                }
                if n == 0 {
                    out.push(b'\\');
                    out.push(b'x');
                } else {
                    out.push((val & 0xff) as u8);
                }
            }
            other => {
                out.push(b'\\');
                out.push(other);
                i += 2;
            }
        }
    }
    out
}
